Lib/inspect.py

```diff
@@ -792,25 +792,57 @@ def findsource(object):
         return lines, 0
 
     if isclass(object):
-        name = object.__name__
-        pat = re.compile(r'^(\s*)class\s*' + name + r'\b')
-        # make some effort to find the best matching class definition:
-        # use the one with the least indentation, which is the one
-        # that's most probably not inside a function definition.
-        candidates = []
-        for i in range(len(lines)):
-            match = pat.match(lines[i])
-            if match:
-                # if it's at toplevel, it's already the best one
-                if lines[i][0] == 'c':
-                    return lines, i
-                # else add whitespace to candidate list
-                candidates.append((match.group(1), i))
-        if candidates:
-            # this will sort by whitespace, and by line number,
-            # less whitespace first
-            candidates.sort()
-            return lines, candidates[0][1]
+        import ast
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def __init__(self, source, qualname, *args, **kwargs):
+                self.stack = []
+                self.source = source
+                self.line_number = None
+                self.qualname = qualname
+                super().__init__(*args, **kwargs)
+
+            def visit_ClassDef(self, node):
+                set_top_level_class = False
+
+                # Push the node on stack when there is a module level class
+                if not self.stack:
+                    set_top_level_class = True
+                    self.stack.append(node.name)
+
+                qualname = '.'.join(self.stack)
```

Comment:
Just inline it.

Reply:
I feel this is more readable that I am joining the stack to form qualname than `if self.qualname == '.'join(self.stack)` . Let me know if it needs to be inlined.

Reply:
It is needed if use local variables from the outer scope. Then the condition will be `if qualname == '.'join(stack)`.